From patchwork Tue Sep 17 18:57:52 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Marc Orr <marcorr@google.com>
X-Patchwork-Id: 11149333
Return-Path: <SRS0=dK0e=XM=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 3293816B1
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 17 Sep 2019 18:58:00 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 0AEA021670
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 17 Sep 2019 18:58:00 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=google.com header.i=@google.com
 header.b="dgfCsN8c"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728238AbfIQS57 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 17 Sep 2019 14:57:59 -0400
Received: from mail-pg1-f202.google.com ([209.85.215.202]:54254 "EHLO
        mail-pg1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725927AbfIQS56 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 17 Sep 2019 14:57:58 -0400
Received: by mail-pg1-f202.google.com with SMTP id j9so2563616pgk.20
        for <kvm@vger.kernel.org>; Tue, 17 Sep 2019 11:57:58 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=4S7/wfUBRRdORQOL8HzLDNrFII6NZGP9NSs87iEQGJI=;
        b=dgfCsN8cXfYnMLyzDSHD5sRPiVf/J13oZ/+EDolHwLnb6vq/O7xePnl1T8Q6gvgq5c
         eqRoHxCkus+ct8GtS52CofjODKVIv3sq3uVLuOUc0vDG+lYJI6MWUOEbfwJhaaZCc5P2
         EptVTB652w8K9dL5DmNzGb5nhp1K0S3jKvOrI41YmzRhvchfWAqFhpk18jZc7XmUrD8n
         g/K+nJWkhjwcD6qHNEktWt0XpZM9oF8nJVecEGP9pSXz+6sSp7bawYJdLG/mC+RxxDS3
         obQzEzx1wex1SPKeVDBIPAz5aY5Yx8I4kSnz0yvI1Cvk3WS1iwTFAJh+qHGNmj9u5Q4U
         odYg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=4S7/wfUBRRdORQOL8HzLDNrFII6NZGP9NSs87iEQGJI=;
        b=phhEJWAmcYzSdmX0pW82ZimdBWNpNL/uoeSqPAKiiYtqLLYepeH08TrT1XteHvnvwa
         PoHnj/lbuUhZnUu2Hnw7vp5ngXYNdlX4JIfpxT+OtEvIOqsoSHydFc5sArAOyJ/H/gQW
         v4x1hNUhotE29Msun1taM2p1/RGDsyHirKOV1cXyKouE8d9EASXzqw3hgbXd4J6jDpl3
         ekqYzipFPu/Y0gkY5cXwnkYUr9Q7I8Bo87qQorJP/swxSQgUZfARK99jXbKOWksV0GNf
         raAv4jWEYSHKeQbxU3WsWNyWzv9PVULpVmkt9Nv9N7twTcUDYte7QAWCfR27aleenNLP
         gmSQ==
X-Gm-Message-State: APjAAAUA+opERjhYNj17/uIIMihg4oHpLjfm1uKvR9UtyEe2b68+hMPS
        nWEb0GDE1SVYJI97r90vzveQ2VTDCWpXE0elM7DOCVOav0XYAivn1N6ZSsVhdOX+c0v3Mw+E+Cd
        gOfL9/DT+2/mSkNxZN8tV29PaSvtynhpbfuSq3gUff7MEn4pfp0Js7qoKy8CD
X-Google-Smtp-Source: 
 APXvYqzKuoJV5x6zfHavf5YI3H7e/wI8+fl8XKqlyhyLaY4M3D44ha3MNDbzFoqUNiDLE6YRU0JgT7QP17Mx
X-Received: by 2002:a65:6294:: with SMTP id f20mr285699pgv.349.1568746677809;
 Tue, 17 Sep 2019 11:57:57 -0700 (PDT)
Date: Tue, 17 Sep 2019 11:57:52 -0700
Message-Id: <20190917185753.256039-1-marcorr@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.23.0.237.gc6a4ce50a0-goog
Subject: [kvm-unit-tests PATCH v3 1/2] x86: nvmx: fix bug in __enter_guest()
From: Marc Orr <marcorr@google.com>
To: kvm@vger.kernel.org, jmattson@google.com, pshier@google.com,
        sean.j.christopherson@intel.com, krish.sadhukhan@oracle.com
Cc: Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

__enter_guest() should only set the launched flag when a launch has
succeeded. Thus, don't set the launched flag when the VMX_ENTRY_FAILURE,
bit 31, is set in the VMCS exit reason.

Signed-off-by: Marc Orr <marcorr@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/vmx.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/x86/vmx.c b/x86/vmx.c
index 6079420db33a..7313c78f15c2 100644
--- a/x86/vmx.c
+++ b/x86/vmx.c
@@ -1820,7 +1820,7 @@ static void __enter_guest(u8 abort_flag, struct vmentry_failure *failure)
 		abort();
 	}
 
-	if (!failure->early) {
+	if (!failure->early && !(vmcs_read(EXI_REASON) & VMX_ENTRY_FAILURE)) {
 		launched = 1;
 		check_for_guest_termination();
 	}

From patchwork Tue Sep 17 18:57:53 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Marc Orr <marcorr@google.com>
X-Patchwork-Id: 11149335
Return-Path: <SRS0=dK0e=XM=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 2C5181747
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 17 Sep 2019 18:58:07 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 007B1214AF
	for <patchwork-kvm@patchwork.kernel.org>;
 Tue, 17 Sep 2019 18:58:06 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=google.com header.i=@google.com
 header.b="vKfz755p"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729722AbfIQS6G (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 17 Sep 2019 14:58:06 -0400
Received: from mail-pl1-f202.google.com ([209.85.214.202]:50053 "EHLO
        mail-pl1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1729594AbfIQS6G (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 17 Sep 2019 14:58:06 -0400
Received: by mail-pl1-f202.google.com with SMTP id p8so2642932plo.16
        for <kvm@vger.kernel.org>; Tue, 17 Sep 2019 11:58:04 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=QZu+4189/LnGT4oApwl9NR7yKk0T4xKQPb7UW0581s4=;
        b=vKfz755plDfYBXp18kYwMTSe1aRE8kFkL11xbjwRvVrzBpwCpac6vON2TyTospoPE9
         TXDQipA3Mco0egbMsf5xW5sYhvyaSKrCasdi29rhlrkyDb7JSvxIVajbSwK5vW/EOCAq
         AStqOrQCa0lFKCvtoUcwQWJe7mnFG+56uGWfQvLTBv7HJBCXnOeEQyQPkaguZXUoF5Im
         AHWoYN78NkI5k71V1WAla+F+QlOfS8wS0d4akNpjtsUOmvkUzt4XSYKWcQvXiZbbbZO9
         OSe63e48i5eJabwTZFMdKrAGZNydw15wymV0YtFBhR7Jr16r7nCrcjCDO4t6LvsBgrCl
         ciqw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=QZu+4189/LnGT4oApwl9NR7yKk0T4xKQPb7UW0581s4=;
        b=kEhTNiTcq170i7C2cce3yA7ex1bKzl0VXY07OgwsKSebdTZS6lbj9BRuN2yVfOGyGy
         jf509Lw/gGvxhUiQs3lB1qOr9BreyAA1vOSKXcU4kB5pB6jKnQqOC1TtRO+iF/SY6k5l
         2XPtA5Nrk6TJBbh7h+8H/BHMW3b1mb0n+3Evwr4J6C5zA8y4jl+pQps8XPD/rlQwXtMC
         tsfFYLzawtbJXwhA8IvtBM/4yAr0vSiYZ1tWl+kdKMswMQcrFJfpBvcOMvG8QW7Bxj9p
         myEbBZO54uNLMPpOfz3m+jxEYEX8TSzynzSxQNMtgucDAc5ZJHdvkoL5Y2fOeMhDp59Q
         KEFw==
X-Gm-Message-State: APjAAAVfU5vTUSqD/vXihtR8kzTxXcXIzmfcBGb6yqj7ahub7gzTw1Pq
        IZDoVx59e1eo0/mVAEk81U+cX3KsZLJygEooAJDvAUsOt9nQTpPdiNTfIIKokLToh2pN42KInRY
        JNWlmZs2QzMpAgneNhbs4kmGB/MCmiXzQ/EgnxGTcwaOEuxbT1powHm55KShh
X-Google-Smtp-Source: 
 APXvYqxeSF6O2LySjKZf7parLX8n7WNbLKQKl8g3iF5WetMTZ+zzSUORha5kQ5L1IK+4NouP7396RHhlux5M
X-Received: by 2002:a65:450a:: with SMTP id n10mr297924pgq.432.1568746683246;
 Tue, 17 Sep 2019 11:58:03 -0700 (PDT)
Date: Tue, 17 Sep 2019 11:57:53 -0700
In-Reply-To: <20190917185753.256039-1-marcorr@google.com>
Message-Id: <20190917185753.256039-2-marcorr@google.com>
Mime-Version: 1.0
References: <20190917185753.256039-1-marcorr@google.com>
X-Mailer: git-send-email 2.23.0.237.gc6a4ce50a0-goog
Subject: [kvm-unit-tests PATCH v3 2/2] x86: nvmx: test max atomic switch MSRs
From: Marc Orr <marcorr@google.com>
To: kvm@vger.kernel.org, jmattson@google.com, pshier@google.com,
        sean.j.christopherson@intel.com, krish.sadhukhan@oracle.com
Cc: Marc Orr <marcorr@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Excerise nested VMX's atomic MSR switch code (e.g., VM-entry MSR-load
list) at the maximum number of MSRs supported, as described in the SDM,
in the appendix chapter titled "MISCELLANEOUS DATA".

Suggested-by: Jim Mattson <jmattson@google.com>
Signed-off-by: Marc Orr <marcorr@google.com>
---
 lib/alloc_page.c  |   5 ++
 lib/alloc_page.h  |   1 +
 x86/unittests.cfg |   2 +-
 x86/vmx_tests.c   | 131 ++++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 138 insertions(+), 1 deletion(-)

diff --git a/lib/alloc_page.c b/lib/alloc_page.c
index 97d13395ff08..ed236389537e 100644
--- a/lib/alloc_page.c
+++ b/lib/alloc_page.c
@@ -53,6 +53,11 @@ void free_pages(void *mem, unsigned long size)
 	spin_unlock(&lock);
 }
 
+void free_pages_by_order(void *mem, unsigned long order)
+{
+	free_pages(mem, 1ul << (order + PAGE_SHIFT));
+}
+
 void *alloc_page()
 {
 	void *p;
diff --git a/lib/alloc_page.h b/lib/alloc_page.h
index 5cdfec57a0a8..739a91def979 100644
--- a/lib/alloc_page.h
+++ b/lib/alloc_page.h
@@ -14,5 +14,6 @@ void *alloc_page(void);
 void *alloc_pages(unsigned long order);
 void free_page(void *page);
 void free_pages(void *mem, unsigned long size);
+void free_pages_by_order(void *mem, unsigned long order);
 
 #endif
diff --git a/x86/unittests.cfg b/x86/unittests.cfg
index 694ee3d42f3a..05122cf91ea1 100644
--- a/x86/unittests.cfg
+++ b/x86/unittests.cfg
@@ -227,7 +227,7 @@ extra_params = -cpu qemu64,+umip
 
 [vmx]
 file = vmx.flat
-extra_params = -cpu host,+vmx -append "-exit_monitor_from_l2_test -ept_access* -vmx_smp* -vmx_vmcs_shadow_test"
+extra_params = -cpu host,+vmx -append "-exit_monitor_from_l2_test -ept_access* -vmx_smp* -vmx_vmcs_shadow_test -atomic_switch_overflow_msrs_test"
 arch = x86_64
 groups = vmx
 
diff --git a/x86/vmx_tests.c b/x86/vmx_tests.c
index f035f24a771a..fb665f38b1e5 100644
--- a/x86/vmx_tests.c
+++ b/x86/vmx_tests.c
@@ -8570,6 +8570,134 @@ static int invalid_msr_entry_failure(struct vmentry_failure *failure)
 	return VMX_TEST_VMEXIT;
 }
 
+/*
+ * The max number of MSRs in an atomic switch MSR list is:
+ * (111B + 1) * 512 = 4096
+ *
+ * Each list entry consumes:
+ * 4-byte MSR index + 4 bytes reserved + 8-byte data = 16 bytes
+ *
+ * Allocate 128 kB to cover max_msr_list_size (i.e., 64 kB) and then some.
+ */
+static const u32 msr_list_page_order = 5;
+
+static void atomic_switch_msr_limit_test_guest(void)
+{
+	vmcall();
+}
+
+static void populate_msr_list(struct vmx_msr_entry *msr_list,
+			      size_t byte_capacity, int count)
+{
+	int i;
+
+	for (i = 0; i < count; i++) {
+		msr_list[i].index = MSR_IA32_TSC;
+		msr_list[i].reserved = 0;
+		msr_list[i].value = 0x1234567890abcdef;
+	}
+
+	memset(msr_list + count, 0xff,
+	       byte_capacity - count * sizeof(*msr_list));
+}
+
+static int max_msr_list_size(void)
+{
+	u32 vmx_misc = rdmsr(MSR_IA32_VMX_MISC);
+	u32 factor = ((vmx_misc & GENMASK(27, 25)) >> 25) + 1;
+
+	return factor * 512;
+}
+
+static void atomic_switch_msrs_test(int count)
+{
+	struct vmx_msr_entry *vm_enter_load;
+        struct vmx_msr_entry *vm_exit_load;
+        struct vmx_msr_entry *vm_exit_store;
+	int max_allowed = max_msr_list_size();
+	int byte_capacity = 1ul << (msr_list_page_order + PAGE_SHIFT);
+	/* Exceeding the max MSR list size at exit trigers KVM to abort. */
+	int exit_count = count > max_allowed ? max_allowed : count;
+	int cleanup_count = count > max_allowed ? 2 : 1;
+	int i;
+
+	/*
+	 * Check for the IA32_TSC MSR,
+	 * available with the "TSC flag" and used to populate the MSR lists.
+	 */
+	if (!(cpuid(1).d & (1 << 4))) {
+		report_skip(__func__);
+		return;
+	}
+
+	/* Set L2 guest. */
+	test_set_guest(atomic_switch_msr_limit_test_guest);
+
+	/* Setup atomic MSR switch lists. */
+	vm_enter_load = alloc_pages(msr_list_page_order);
+	vm_exit_load = alloc_pages(msr_list_page_order);
+	vm_exit_store = alloc_pages(msr_list_page_order);
+
+	vmcs_write(ENTER_MSR_LD_ADDR, (u64)vm_enter_load);
+	vmcs_write(EXIT_MSR_LD_ADDR, (u64)vm_exit_load);
+	vmcs_write(EXIT_MSR_ST_ADDR, (u64)vm_exit_store);
+
+	/*
+	 * VM-Enter should succeed up to the max number of MSRs per list, and
+	 * should not consume junk beyond the last entry.
+	 */
+	populate_msr_list(vm_enter_load, byte_capacity, count);
+	populate_msr_list(vm_exit_load, byte_capacity, exit_count);
+	populate_msr_list(vm_exit_store, byte_capacity, exit_count);
+
+	vmcs_write(ENT_MSR_LD_CNT, count);
+	vmcs_write(EXI_MSR_LD_CNT, exit_count);
+	vmcs_write(EXI_MSR_ST_CNT, exit_count);
+
+	if (count <= max_allowed) {
+		enter_guest();
+		assert_exit_reason(VMX_VMCALL);
+		skip_exit_vmcall();
+	} else {
+		u32 exit_reason;
+		u32 exit_reason_want;
+		u32 exit_qual;
+
+		enter_guest_with_invalid_guest_state();
+
+		exit_reason = vmcs_read(EXI_REASON);
+		exit_reason_want = VMX_FAIL_MSR | VMX_ENTRY_FAILURE;
+		report("exit_reason, %u, is %u.",
+		       exit_reason == exit_reason_want, exit_reason,
+		       exit_reason_want);
+
+		exit_qual = vmcs_read(EXI_QUALIFICATION);
+		report("exit_qual, %u, is %u.", exit_qual == max_allowed + 1,
+		       exit_qual, max_allowed + 1);
+	}
+
+	/* Cleanup. */
+	vmcs_write(ENT_MSR_LD_CNT, 0);
+	vmcs_write(EXI_MSR_LD_CNT, 0);
+	vmcs_write(EXI_MSR_ST_CNT, 0);
+	for (i = 0; i < cleanup_count; i++) {
+		enter_guest();
+		skip_exit_vmcall();
+	}
+	free_pages_by_order(vm_enter_load, msr_list_page_order);
+	free_pages_by_order(vm_exit_load, msr_list_page_order);
+	free_pages_by_order(vm_exit_store, msr_list_page_order);
+}
+
+static void atomic_switch_max_msrs_test(void)
+{
+	atomic_switch_msrs_test(max_msr_list_size());
+}
+
+static void atomic_switch_overflow_msrs_test(void)
+{
+	atomic_switch_msrs_test(max_msr_list_size() + 1);
+}
 
 #define TEST(name) { #name, .v2 = name }
 
@@ -8660,5 +8788,8 @@ struct vmx_test vmx_tests[] = {
 	TEST(ept_access_test_paddr_read_execute_ad_enabled),
 	TEST(ept_access_test_paddr_not_present_page_fault),
 	TEST(ept_access_test_force_2m_page),
+	/* Atomic MSR switch tests. */
+	TEST(atomic_switch_max_msrs_test),
+	TEST(atomic_switch_overflow_msrs_test),
 	{ NULL, NULL, NULL, NULL, NULL, {0} },
 };
